lib/compiler/aro_translate_c.zig

```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+    }
+
+    var is_unnamed = false;
+    var name = bare_name;
+    if (c.unnamed_typedefs.get(@intFromPtr(record_decl))) |typedef_name| {
+        bare_name = typedef_name;
+        name = typedef_name;
+    } else {
+        // Record declarations such as `struct {...} x` have no name but they're not
+        // anonymous hence here isAnonymousStructOrUnion is not needed
+        if (ty.isAnonymousRecord(c.comp)) {
+            bare_name = try std.fmt.allocPrint(c.arena, "unnamed_{d}", .{c.getMangle()});
+            is_unnamed = true;
+        }
+        name = try std.fmt.allocPrint(c.arena, "{s}_{s}", .{ container_kind_name, bare_name });
+        if (toplevel and !is_unnamed) {
+            name = try mangleWeakGlobalName(c, name);
+        }
+    }
+    if (!toplevel) name = try bs.makeMangledName(c, name);
+    try c.decl_table.putNoClobber(c.gpa, @intFromPtr(record_decl), name);
+
+    const is_pub = toplevel and !is_unnamed;
+    const init_node = blk: {
+        var fields = try std.ArrayList(ast.Payload.Record.Field).initCapacity(c.gpa, record_decl.fields.len);
+        defer fields.deinit();
+
+        var functions = std.ArrayList(ZigNode).init(c.gpa);
+        defer functions.deinit();
+
+        var unnamed_field_count: u32 = 0;
+
+        for (record_decl.fields, field_nodes, 0..) |field, field_node, i| {
+            const field_raw_ty = node_types[@intFromEnum(field_node)];
+            const field_qt = field_raw_ty.canonicalize(.preserve_quals);
+            const field_loc = field.name_tok;
+
+            // Demote record to opaque if it contains a bitfield
+            if (!field.isRegularField()) {
+                try c.opaque_demotes.put(c.gpa, @intFromPtr(record_decl), {});
+                try warn(c, scope, field_loc, "{s} demoted to opaque type - has bitfield", .{container_kind_name});
+                break :blk ZigTag.opaque_literal.init();
+            }
+
+            var field_name = c.mapper.lookup(field.name);
+            if (!field.isNamed()) {
+                field_name = try std.fmt.allocPrint(c.arena, "unnamed_{d}", .{unnamed_field_count});
+                unnamed_field_count += 1;
+            }
+            // TODO: Add support for flexible array fields
+            if (field_qt.isEnumOrRecord()) try transDecl(c, scope, field_node);
+            const field_type = transType(c, scope, field_qt, .preserve_quals, field_loc) catch |err| switch (err) {
+                error.UnsupportedType => {
+                    try c.opaque_demotes.put(c.gpa, @intFromPtr(ty.canonicalize(.standard).data.record), {});
+                    try warn(c, scope, 0, "{s} demoted to opaque type - unable to translate type of field {s}", .{ container_kind_name, field_name });
+                    break :blk ZigTag.opaque_literal.init();
+                },
+                else => |e| return e,
+            };
+
+            var field_ty_with_attrs = field.ty;
```

Comment:
We compute record layouts so I don't think this is needed; `field.layout` tells us the field's offset in bits from the start of the record. Also note that the requested alignment is not always granted by the compiler, but the record layout code accounts for that.

Reply:
I see, that makes sense. Getting the requested attribute isn't the way to go then. I'm not sure we can get the alignment from just the field offset though. For example, how would this test case work? The offset is 0 with or without the attribute. Would we use one of the values in `TypeLayout`?

```c
// struct_with_aligned_field.c
struct foo {
    __attribute__((aligned(1))) short bar;
};
```

```zig
// translated zig code
pub const struct_foo = extern struct {
    bar: c_short align(1) = @import("std").mem.zeroes(c_short),
};
```

Reply:
Sorry, should have mentioned the record's `type_layout` as well. The example you posted seems like a bug in current translate-c - in C the `aligned` attribute can only increase a field's alignment, not decrease it (unless the record is packed).

```c
// x86_64-linux
struct foo {
    __attribute__((aligned(1))) short bar;
};
_Static_assert(_Alignof(struct foo) == 2, "");
```
whereas the translated zig struct you pasted will have an alignment of 1.

I believe you want the `pointer_alignment_bits` of the record's `type_layout`, which tells you the required alignment of valid pointers to that type.

Reply:
Okay, that makes sense! I'll play around with those fields and see what I can come up with. 

I forgot that the aligned attribute can only increase the alignment. I think that's a bug too. The clang translate-c is only looking at the attributes, not the actual record layout. I'll extract this and make a tracking issue.